### ISSUE TITLE
Add support of more resources

### ISSUE THREAD

**Issue by davidebianchi** [kube-green/kube-green#77]
What type of resources should be handled by kube-green?

* [ ] traefik ingress route
* [ ] Replica Sets
* [ ] StatefulSet

**Comment by jwcesign** [comment]
Can we also delete node and restore it for some public cloud, like gcp?

**Comment by davidebianchi** [comment]
>  Is it possible to use Kube-green with replica sets, daemon sets etc so the nodes could scale down to zero?

For replica sets yes, we could add the support. For the daemon, there will be one per node so I don't think it's ok to remove programmatically. If the cluster autoscaler is enabled, the node should downscale automatically. If it is not, you should modify the deamon accordingly. If you have any problem about this, you can see this docs [to troubleshooting the nodes downscaling](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#i-have-a-couple-of-nodes-with-low-utilization-but-they-are-not-scaled-down-why)

> Can we also delete node and restore it for some public cloud, like gcp?

Yes, it is possible to do it with the node autoscaler on gcp. The previous link could help you?

**Comment by jgerber-sbb** [comment]
Support for DeploymentConfig (OpenShift specific) would be nice:
https://docs.openshift.com/container-platform/latest/applications/deployments/what-deployments-are.html
https://docs.openshift.com/container-platform/latest/rest_api/workloads_apis/deploymentconfig-apps-openshift-io-v1.html

**Comment by djmcgreal-cc** [comment]
+StatefulSets? I'd like to make a PR or so, but want to gauge whether it's already been thought of but there's a more fundamental difficulty in the way?

I'd like to deploy this to our pre-production clusters but in our case, there's not much point unless we can take the node down.
Thanks for this!

**Comment by djmcgreal-cc** [comment]
And possibly a some support for arbitrary configuration for CRDs in the SleepInfo, like a selector and strategic/merge/json patch?

**Comment by davidebianchi** [comment]
Hi @djmcgreal-cc! Sure, we could add the StatefulSets. I haven't looked yet for this resource in the detail, but it has the replicas configuration if I well remember, so I suppose it could be add easily.
Do you have some nodes with only some StatefulSets?

> And possibly a some support for arbitrary configuration for CRDs in the SleepInfo, like a selector and strategic/merge/json patch?
What do you mean with arbitrary configuration? Can you add more details on the expected use case or some more description on what can be done?

**Comment by djmcgreal** [comment]
Precisely. So that the project doesn’t need to track all CRDs in the ecosystem

**Comment by potitoaghilar** [comment]
> Hi @djmcgreal-cc! Sure, we could add the StatefulSets. I haven't looked yet for this resource in the detail, but it has the replicas configuration if I well remember, so I suppose it could be add easily. Do you have some nodes with only some StatefulSets?
> 
> > And possibly a some support for arbitrary configuration for CRDs in the SleepInfo, like a selector and strategic/merge/json patch?
> 
> What do you mean with arbitrary configuration? Can you add more details on the expected use case or some more description on what can be done? Maybe a generic way to describe which resource should be managed and what are the changes which should be done to disable the specific resource?

We are also interested in StatefulSet scaling to zero. In our case we have StatefulSets and Deployments on same node.

**Comment by bygui86** [comment]
Maybe it could be useful a catalog of CRDs...
https://github.com/datreeio/CRDs-catalog

**Comment by davidebianchi** [comment]
Hi all, I'm thinking about something like json patch to try to generalize the support on various resources.
So, something like:

```yaml
  patchesJson6902:
    - target:
        version: "apps/v1"
        kind: Deployment
        # labelSelector: "env=dev"
      patch:
        key: "#/spec/replicas"
        value: 0
    - target:
        apiVersion: "batch/v1"
        kind: CronJob
      patch:
        key: "#/spec/suspend"
        value: true
```

In this way, we could also add some first class resources (like deployment and cronjobs) which are supported by default, and the user could also add the patch for a custom CRD which want to manage.
What do you think about? Any hint on this?

**Comment by djmcgreal-cc** [comment]
Sounds totally sensible to me

**Comment by jgpelaez** [comment]
@davidebianchi It totally makes sense the approach for doing it agnostic to the CRD, as for example there will be some operator that can have in a CRD the number of replicas that manages later a deployment or statefulset

**Comment by davidebianchi** [comment]
I created a PR to make a POC of this configuration (which I partially changed). The POC seems effective, the next step is to make it working e2e and start testing with some new configurations. Feedback are as always welcome! 😄

**Comment by davidebianchi** [comment]
Support of all resources is now released in [v0.6.0-rc.1](https://github.com/kube-green/kube-green/releases/tag/v0.6.0-rc.1) and it is possible to test.
The support of StatefulSet by default is added in the same version by #425 

Close this, it will be released in stable version in v0.6.0
